solidity/test/bridge/VendingMachine.test.js

```diff
@@ -0,0 +1,604 @@
+const { expect } = require("chai")
+const {
+  increaseTime,
+  lastBlockTime,
+  to1e18,
+  to1ePrecision,
+  ZERO_ADDRESS,
+} = require("../helpers/contract-test-helpers")
+
+describe("VendingMachine", () => {
+  let tbtcV1
+  let tbtcV2
+  let vendingMachine
+
+  let governance
+  let tokenHolder
+  let thirdParty
+
+  const unmintFee = to1ePrecision(1, 15) // 0.001
+  const initialBalance = to1e18(5) // 5 TBTC v1
+
+  beforeEach(async () => {
+    ;[
+      deployer,
+      unmintFeeUpdateInitiator,
+      vendingMachineUpdateInitiator,
+      governance,
+      tokenHolder,
+      thirdParty,
+    ] = await ethers.getSigners()
+
+    const TestERC20 = await ethers.getContractFactory("TestERC20")
+    tbtcV1 = await TestERC20.deploy()
+    await tbtcV1.deployed()
+
+    const TBTCToken = await ethers.getContractFactory("TBTCToken")
+    tbtcV2 = await TBTCToken.deploy()
+    await tbtcV2.deployed()
+
+    await tbtcV1.mint(tokenHolder.address, initialBalance)
+
+    const VendingMachine = await ethers.getContractFactory("VendingMachine")
+    vendingMachine = await VendingMachine.deploy(
+      tbtcV1.address,
+      tbtcV2.address,
+      unmintFee
+    )
+    await vendingMachine.deployed()
+
+    await vendingMachine.connect(deployer).transferOwnership(governance.address)
+    await vendingMachine
+      .connect(deployer)
+      .transferUnmintFeeUpdateInitiatorRole(unmintFeeUpdateInitiator.address)
+    await vendingMachine
+      .connect(deployer)
+      .transferVendingMachineUpdateInitiatorRole(
+        vendingMachineUpdateInitiator.address
+      )
+    await tbtcV2.connect(deployer).transferOwnership(vendingMachine.address)
+
+    await tbtcV1
+      .connect(tokenHolder)
+      .approve(vendingMachine.address, initialBalance)
+  })
+
+  describe("deposit", () => {
+    describe("when TBTC v1 owner has not enough tokens", () => {
+      it("should revert", async () => {
+        const amount = initialBalance.add(1)
+        await tbtcV1
+          .connect(tokenHolder)
+          .approve(vendingMachine.address, amount)
+        await expect(
+          vendingMachine.connect(tokenHolder).mint(amount)
+        ).to.be.revertedWith("Transfer amount exceeds balance")
+      })
+    })
+
+    describe("when TBTC v1 owner has enough tokens", () => {
+      let tx
+
+      beforeEach(async () => {
+        tx = await vendingMachine.connect(tokenHolder).mint(initialBalance)
+      })
+
+      it("should mint the same amount of TBTC v2", async () => {
+        expect(await tbtcV2.balanceOf(tokenHolder.address)).is.equal(
+          initialBalance
```

Comment:
That makes sense but only if the allowance stays at `initialBalance` which is I think what you meant. Otherwise, we are testing the token. So:
- allowance at `initialBalance`, minting `initialBalance - 100`
- allowance at `initialBalance`, minting `initialBalance`

Reply:
ed5743e8c2ad60e441c1b8813aee88604bed79ea